File: emp_importance.py
// Time: O(n) where n is no. of employees
// Space: O(n)
// Passed on Leetcode: Yes
// Approach: Use a queue(BFS) to iterate through given employees subordinates. acess the subordinates effectively by storing the 
// index of each employee id in employees array for faster access. Keep incrementing the total importance value as we pop elements from queue.

"""
# Definition for Employee.
class Employee(object):
    def __init__(self, id, importance, subordinates):
    	#################
        :type id: int
        :type importance: int
        :type subordinates: List[int]
        #################
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution(object):
    def getImportance(self, employees, id):
        """
        :type employees: List[Employee]
        :type id: int
        :rtype: int
        """
        
        queue=[]
        res=0
        d={}
        
        for i in range(len(employees)):
            if(employees[i].id not in d):
                d[employees[i].id] = i
        
        queue.append(id)
        
        while(queue):
            
            popped=queue.pop(0)
            res+= employees[d[popped]].importance
            subor= employees[d[popped]].subordinates
            
            for i in range(len(subor)):
                queue.append(subor[i])
        return(res)
            
            
        

File: EmpImportance.java
//BFS approach
// Time Complexity : O(v+e), v = e = n , no. of employees so O(n)
// Space Complexity : O(v+e), v = e = n , no. of employees so O(n) will be stored in hashmap
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english : It is graph represented as a n-ary tree as an employee will not report to multiple
// managers. Maintain a hashmap to store fetch the employee in O(1) and do bfs adding importance of each employee in queue.

/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        Map<Integer,Employee> map = new HashMap<>();
        for(Employee emp : employees){
            map.put(emp.id, emp);
        }

        Employee e = map.get(id);
        Queue<Employee> q = new LinkedList<>();
        q.add(e);
        int imp = 0;

        while(!q.isEmpty()){
            e = q.poll();
            imp += e.importance;
            List<Integer> subordinates= e.subordinates;
            for(int sub : subordinates){
                q.add(map.get(sub));
            }
        }

        return imp;
    }
}

//DFS approach
// Time Complexity : O(v+e), v = e = n , no. of employees so O(n)
// Space Complexity : O(v+e), v = e = n , no. of employees so O(n) will be stored in hashmap
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english : It is graph represented as a n-ary tree as an employee will not report to multiple
// managers. Maintain a hashmap to store fetch the employee in O(1) and do dfs adding importance of each subordinates.

/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    Map<Integer, Employee> map;
    int imp = 0;

    public int getImportance(List<Employee> employees, int id) {
        this.map = new HashMap<>();
        for (Employee emp : employees) {
            map.put(emp.id, emp);
        }

        dfs(employees, id);
        return imp;
    }

    private void dfs(List<Employee> employees, int id) {
        //base - no need, for loop will only run for subordinates

        Employee e = map.get(id);
        imp += e.importance;
        List<Integer> subordinates = e.subordinates;

        for (int sid : subordinates) {
            dfs(employees, sid);
        }
    }
}
